### PR TITLE
Fix squeeze() backward in edge case

### LINKED ISSUE #4778: squeeze on 1x1 tensor causes error in gradients while view(-1) doesn't
```python
import torch
from torch.autograd import grad, Variable

w = Variable(torch.ones([3, 1])).cuda()
xyz = Variable(torch.FloatTensor([[0, 1.3, 2], ]), requires_grad=True).cuda()
res = torch.matmul(xyz, w)

# Works
torch.autograd.grad(res.view(-1), xyz, torch.ones_like(res.squeeze()), retain_graph=True)
# Doesn't
torch.autograd.grad(res.squeeze(), xyz, torch.ones_like(res.squeeze()), retain_graph=True)
```

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #4778 

There was a bug where if one calls `tensor.squeeze().backward()` where `tensor` has size `(1, 1)`, `tensor.grad` doesn't match the shape of `tensor`. This is because the result of `tensor.squeeze()` has size `(1,)`, and `squeeze()`'s backwards will `unsqueeze` the grad a number of times equal to the number of dimensions in `tensor` that have size 1.

cc @gchanan -- the code should work with both scalars enabled and disabled because there's a check for `dim() == 1`.

### Test Plan
New unit test for this case

### CODE REVIEW

**Review by gchanan — APPROVED** [review]

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -302,6 +302,19 @@ Tensor cumsum_backward(const Tensor & x, int64_t dim) {
...
   int64_t nDims = sizes.size();
+
+  // Let's say the input had size (1, 1). input.squeeze(), with scalars
+  // disabled, produces a result of size (1,). This needs some
+  // special handling because for all other cases we unsqueeze every
+  // dimension that has size 1; doing that here would lead to one extra
+  // unsqueezed dimension
+  if (self.sizes().equals({1})) {
+    for (int64_t dim = 0; dim < nDims - 1; dim++) {
+      result = result.unsqueeze(dim);
```
Comment (apaszke) [inline tools/autograd/templates/Functions.cpp]: It would be nicer to use `view` for this - every unsqueeze in here is an autograd op, so you end up doing 5x the work (and the graph gets larger).
